va/tlsalpn.go

```diff
@@ -150,36 +184,69 @@ func (va *ValidationAuthorityImpl) getChallengeCert(
 	cs := conn.(*tls.Conn).ConnectionState()
 	certs := cs.PeerCertificates
 	if len(certs) == 0 {
-		va.log.Infof("%s challenge for %s resulted in no certificates", core.ChallengeTypeTLSALPN01, identifier.Value)
+		va.log.Infof("%s challenge for %s resulted in no certificates", core.ChallengeTypeTLSALPN01, ident.Value)
 		return nil, nil, berrors.UnauthorizedError("No certs presented for %s challenge", core.ChallengeTypeTLSALPN01)
 	}
 	for i, cert := range certs {
 		va.log.AuditInfof("%s challenge for %s received certificate (%d of %d): cert=[%s]",
-			core.ChallengeTypeTLSALPN01, identifier.Value, i+1, len(certs), hex.EncodeToString(cert.Raw))
+			core.ChallengeTypeTLSALPN01, ident.Value, i+1, len(certs), hex.EncodeToString(cert.Raw))
 	}
 	return certs[0], &cs, nil
 }
 
-func checkExpectedSAN(cert *x509.Certificate, name identifier.ACMEIdentifier) error {
-	if len(cert.DNSNames) != 1 {
-		return errors.New("wrong number of dNSNames")
+func checkExpectedSAN(cert *x509.Certificate, ident identifier.ACMEIdentifier) error {
+	var expectedSANBytes []byte
+	switch ident.Type {
+	case identifier.TypeDNS:
+		if len(cert.DNSNames) != 1 || len(cert.IPAddresses) != 0 {
+			return errors.New("wrong number of identifiers")
+		}
+		if !strings.EqualFold(cert.DNSNames[0], ident.Value) {
+			return errors.New("identifier does not match expected identifier")
+		}
+		bytes, err := asn1.Marshal([]asn1.RawValue{
+			{Tag: 2, Class: 2, Bytes: []byte(ident.Value)},
+		})
+		if err != nil {
+			return errors.New("composing SAN extension")
+		}
+		expectedSANBytes = bytes
+	case identifier.TypeIP:
+		if len(cert.IPAddresses) != 1 || len(cert.DNSNames) != 0 {
+			return errors.New("wrong number of identifiers")
+		}
+		if !cert.IPAddresses[0].Equal(net.ParseIP(ident.Value)) {
+			return errors.New("identifier does not match expected identifier")
+		}
+		netipAddr, err := netip.ParseAddr(ident.Value)
+		if err != nil {
+			return errors.New("parsing IP address identifier")
+		}
+		netipBytes, err := netipAddr.MarshalBinary()
+		if err != nil {
+			return errors.New("marshalling IP address identifier")
+		}
+		bytes, err := asn1.Marshal([]asn1.RawValue{
+			{Tag: 7, Class: 2, Bytes: netipBytes},
+		})
+		if err != nil {
+			return errors.New("composing SAN extension")
+		}
+		expectedSANBytes = bytes
+	default:
+		// This should never happen. The calling function should check the
+		// identifier type.
+		return fmt.Errorf("unknown identifier type: %s", ident.Type)
 	}
 
 	for _, ext := range cert.Extensions {
 		if IdCeSubjectAltName.Equal(ext.Id) {
-			expectedSANs, err := asn1.Marshal([]asn1.RawValue{
-				{Tag: 2, Class: 2, Bytes: []byte(cert.DNSNames[0])},
-			})
-			if err != nil || !bytes.Equal(expectedSANs, ext.Value) {
+			if !bytes.Equal(expectedSANBytes, ext.Value) {
```

Comment:
generally always order these (got, want), so
```suggestion
			if !bytes.Equal(ext.Value, expectedSANBytes) {
```